File: orchestrator/executor.py
"""The orchestration engine: walks the dependency graph wave by wave,
enforcing entry/exit gates, approval checkpoints, bounded retries, fallback,
and rollback/safe-stop -- and supports re-planning by re-invoking `run()`
after the caller marks part of the graph stale.

A "wave" is the set of nodes whose dependencies are all satisfied at a given
moment. Nodes in a wave with `parallel_ok=True` execute concurrently in a
thread pool; the next wave cannot start until the whole current wave (serial
and parallel members) has resolved, which is the "synchronization" the
assignment asks for -- e.g. testing and docs both depend on implementation
and run side by side, but release waits for both.
"""
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass

from orchestrator.approvals import Approver
from orchestrator.context import SharedContext
from orchestrator.graph import FAILED, PENDING, ROLLED_BACK, RUNNING, SKIPPED, STALE, SUCCEEDED, Graph, Node
from orchestrator.metrics import RunMetrics, compute_metrics
from orchestrator.observability import EventLog


@dataclass
class RunResult:
    ctx: SharedContext
    graph: Graph
    metrics: RunMetrics
    event_log: EventLog
    outcome: str  # "completed" | "rolled_back" | "safe_stopped"


class Executor:
    def __init__(
        self,
        graph: Graph,
        ctx: SharedContext,
        event_log: EventLog,
        approver: Approver,
        rollback_on_failure: bool = True,
        max_workers: int = 4,
    ):
        graph.validate()
        self.graph = graph
        self.ctx = ctx
        self.event_log = event_log
        self.approver = approver
        self.rollback_on_failure = rollback_on_failure
        self.max_workers = max_workers

    def run(self) -> RunResult:
        is_replan = any(n.status == STALE for n in self.graph.nodes.values())
        self.event_log.emit("replan" if is_replan else "run_started")
        if is_replan:
            self.event_log.emit(
                "run_resumed_after_replan",
                stale_nodes=[n.id for n in self.graph.nodes.values() if n.status == STALE],
            )

        outcome = "completed"
        while True:
            wave = self.graph.ready_nodes()
            if not wave:
                break

            parallel_nodes = [n for n in wave if n.parallel_ok]
            serial_nodes = [n for n in wave if not n.parallel_ok]

            self.event_log.emit(
                "wave_started",
                parallel=[n.id for n in parallel_nodes],
                serial=[n.id for n in serial_nodes],
            )

            if parallel_nodes:
                with ThreadPoolExecutor(max_workers=self.max_workers) as pool:
                    list(pool.map(self._execute_node, parallel_nodes))
            for node in serial_nodes:
                self._execute_node(node)

            failed_this_wave = [n for n in wave if n.status == FAILED]
            if failed_this_wave:
                if self.rollback_on_failure:
                    self._rollback()
                    outcome = "rolled_back"
                else:
                    self._safe_stop()
                    outcome = "safe_stopped"
                break

        self.event_log.emit("run_finished", outcome=outcome)
        metrics = compute_metrics(self.event_log.events, self.graph)
        return RunResult(self.ctx, self.graph, metrics, self.event_log, outcome)

    # -- node execution -----------------------------------------------------

    def _execute_node(self, node: Node) -> None:
        node.status = RUNNING
        self.event_log.emit("node_started", node_id=node.id, stage=node.stage, name=node.name)

        for policy in node.entry_policies:
            pr = policy(self.ctx, node, None)
            if not pr.passed:
                self.ctx.record_policy_violation(node.id, getattr(policy, "__name__", "entry_policy"), pr.message)
                self.event_log.emit(
                    "policy_violation", node_id=node.id, policy=getattr(policy, "__name__", "entry_policy"),
                    gate="entry", message=pr.message,
                )
                node.status = FAILED
                node.last_error = f"entry policy failed: {pr.message}"
                self.event_log.emit("node_failed", node_id=node.id, stage=node.stage, error=node.last_error)
                return

        if node.requires_approval:
            decision = self.approver.decide(self.ctx, node, None)
            self.ctx.record_approval(node.id, decision.approved, decision.approver, decision.rationale)
            self.event_log.emit(
                "approval_decision", node_id=node.id, approved=decision.approved,
                approver=decision.approver, rationale=decision.rationale,
            )
            if not decision.approved:
                node.status = FAILED
                node.last_error = f"approval withheld by {decision.approver}: {decision.rationale}"
                self.event_log.emit("node_failed", node_id=node.id, stage=node.stage, error=node.last_error)
                return

        result = None
        max_attempts = max(1, node.retry.max_attempts)
        for attempt in range(1, max_attempts + 1):
            node.attempts = attempt
            try:
                result = node.run(self.ctx, node)
            except Exception as exc:  # noqa: BLE001 -- executor must not crash on agent errors
                node.last_error = str(exc)
                if attempt < max_attempts:
                    self.event_log.emit(
                        "node_retry", node_id=node.id, attempt=attempt, max_attempts=max_attempts,
                        reason=str(exc),
                    )
                    time.sleep(node.retry.backoff_seconds * attempt)
                    continue
                break

            node.last_result = result  # visible to compensate even if the exit gate rejects this attempt

            exit_failure = None
            for policy in node.exit_policies:
                pr = policy(self.ctx, node, result)
                if not pr.passed:
                    exit_failure = (getattr(policy, "__name__", "exit_policy"), pr.message)
                    break

            if exit_failure is None:
                self.ctx.set_output(node.id, result)
                node.status = SUCCEEDED
                self.event_log.emit("node_succeeded", node_id=node.id, stage=node.stage)
                return

            policy_name, message = exit_failure
            self.ctx.record_policy_violation(node.id, policy_name, message)
            self.event_log.emit(
                "policy_violation", node_id=node.id, policy=policy_name, gate="exit", message=message
            )
            node.last_error = f"exit policy '{policy_name}' failed: {message}"
            if attempt < max_attempts:
                self.event_log.emit(
                    "node_retry", node_id=node.id, attempt=attempt, max_attempts=max_attempts, reason=message
                )
                time.sleep(node.retry.backoff_seconds * attempt)
                continue
            break

        if node.fallback is not None:
            try:
                self.event_log.emit("node_fallback_invoked", node_id=node.id)
                result = node.fallback(self.ctx, node)
                node.last_result = result
                self.ctx.set_output(node.id, result)
                node.status = SUCCEEDED
                self.event_log.emit("node_succeeded", node_id=node.id, stage=node.stage, via_fallback=True)
                return
            except Exception as exc:  # noqa: BLE001
                node.last_error = f"fallback also failed: {exc}"

        node.status = FAILED
        self.event_log.emit("node_failed", node_id=node.id, stage=node.stage, error=node.last_error)

    # -- failure handling -----------------------------------------------------

    def _rollback(self) -> None:
        order = list(reversed(self.graph.topological_order()))
        self.event_log.emit("rollback_started", order=order)
        for nid in order:
            node = self.graph.nodes[nid]
            # A node that failed its own exit gate can still have written side
            # effects before the gate rejected it (last_result is set as soon as
            # node.run() returns, before the exit policy is checked) -- it must
            # be compensated too, not just nodes that fully succeeded.
            if node.status in (SUCCEEDED, FAILED) and node.compensate is not None and node.last_result is not None:
                try:
                    node.compensate(self.ctx, node)
                    was_succeeded = node.status == SUCCEEDED
                    if was_succeeded:
                        node.status = ROLLED_BACK
                    self.event_log.emit("node_rolled_back", node_id=node.id, was_succeeded=was_succeeded)
                except Exception as exc:  # noqa: BLE001
                    self.event_log.emit("node_rollback_failed", node_id=node.id, error=str(exc))
        for node in self.graph.nodes.values():
            if node.status in (PENDING, STALE):
                node.status = SKIPPED
        self.event_log.emit("rollback_finished")

    def _safe_stop(self) -> None:
        skipped = []
        for node in self.graph.nodes.values():
            if node.status in (PENDING, STALE):
                node.status = SKIPPED
                skipped.append(node.id)
        self.event_log.emit(
            "safe_stop",
            skipped_nodes=skipped,
            note="halted without rollback; all previously SUCCEEDED nodes remain in place as last-known-good state",
        )

File: src/main/java/com/example/urlshortener/orchestrator/Executor.java
package com.example.urlshortener.orchestrator;

import java.util.*;
import java.util.concurrent.*;

/**
 * The orchestration engine: walks the dependency graph wave by wave,
 * enforcing entry/exit gates, approval checkpoints, bounded retries,
 * fallback, and rollback/safe-stop -- and supports re-planning by
 * re-invoking {@link #run()} after the caller marks part of the graph
 * stale. Direct, behavior-preserving port of orchestrator/executor.py.
 */
public class Executor {
    private final Graph graph;
    private final SharedContext ctx;
    private final EventLog eventLog;
    private final Approver approver;
    private final boolean rollbackOnFailure;
    private final int maxWorkers;

    public Executor(Graph graph, SharedContext ctx, EventLog eventLog, Approver approver, boolean rollbackOnFailure, int maxWorkers) {
        graph.validate();
        this.graph = graph;
        this.ctx = ctx;
        this.eventLog = eventLog;
        this.approver = approver;
        this.rollbackOnFailure = rollbackOnFailure;
        this.maxWorkers = maxWorkers;
    }

    public Executor(Graph graph, SharedContext ctx, EventLog eventLog, Approver approver, boolean rollbackOnFailure) {
        this(graph, ctx, eventLog, approver, rollbackOnFailure, 4);
    }

    public RunResult run() {
        boolean isReplan = graph.nodes.values().stream().anyMatch(n -> n.status == NodeStatus.STALE);
        eventLog.emit(isReplan ? "replan" : "run_started");
        if (isReplan) {
            List<String> stale = graph.nodes.values().stream().filter(n -> n.status == NodeStatus.STALE).map(n -> n.id).toList();
            eventLog.emit("run_resumed_after_replan", null, Map.of("stale_nodes", stale));
        }

        String outcome = "completed";
        while (true) {
            List<Node> wave = graph.readyNodes();
            if (wave.isEmpty()) break;

            List<Node> parallelNodes = wave.stream().filter(n -> n.parallelOk).toList();
            List<Node> serialNodes = wave.stream().filter(n -> !n.parallelOk).toList();

            eventLog.emit("wave_started", null, Map.of(
                    "parallel", parallelNodes.stream().map(n -> n.id).toList(),
                    "serial", serialNodes.stream().map(n -> n.id).toList()
            ));

            if (!parallelNodes.isEmpty()) {
                ExecutorService pool = Executors.newFixedThreadPool(Math.min(maxWorkers, Math.max(1, parallelNodes.size())));
                try {
                    List<Future<?>> futures = new ArrayList<>();
                    for (Node n : parallelNodes) futures.add(pool.submit(() -> executeNode(n)));
                    for (Future<?> f : futures) f.get();
                } catch (InterruptedException | ExecutionException e) {
                    throw new RuntimeException(e);
                } finally {
                    pool.shutdown();
                }
            }
            for (Node n : serialNodes) executeNode(n);

            List<Node> failedThisWave = wave.stream().filter(n -> n.status == NodeStatus.FAILED).toList();
            if (!failedThisWave.isEmpty()) {
                if (rollbackOnFailure) {
                    rollback();
                    outcome = "rolled_back";
                } else {
                    safeStop();
                    outcome = "safe_stopped";
                }
                break;
            }
        }

        eventLog.emit("run_finished", null, Map.of("outcome", outcome));
        RunMetrics metrics = MetricsCalculator.compute(eventLog.events, graph);
        return new RunResult(ctx, graph, metrics, eventLog, outcome);
    }

    private void executeNode(Node node) {
        node.status = NodeStatus.RUNNING;
        eventLog.emit("node_started", node.id, Map.of("stage", node.stage, "name", node.name));

        for (NamedPolicy policy : node.entryPolicies) {
            PolicyResult pr = policy.check(ctx, node, null);
            if (!pr.passed()) {
                ctx.recordPolicyViolation(node.id, policy.name(), pr.message());
                eventLog.emit("policy_violation", node.id, Map.of("policy", policy.name(), "gate", "entry", "message", pr.message()));
                node.status = NodeStatus.FAILED;
                node.lastError = "entry policy failed: " + pr.message();
                eventLog.emit("node_failed", node.id, Map.of("stage", node.stage, "error", node.lastError));
                return;
            }
        }

        if (node.requiresApproval) {
            ApprovalDecision decision = approver.decide(ctx, node, null);
            ctx.recordApproval(node.id, decision.approved(), decision.approver(), decision.rationale());
            eventLog.emit("approval_decision", node.id, Map.of(
                    "approved", decision.approved(), "approver", decision.approver(), "rationale", decision.rationale()));
            if (!decision.approved()) {
                node.status = NodeStatus.FAILED;
                node.lastError = "approval withheld by " + decision.approver() + ": " + decision.rationale();
                eventLog.emit("node_failed", node.id, Map.of("stage", node.stage, "error", node.lastError));
                return;
            }
        }

        Map<String, Object> result = null;
        int maxAttempts = Math.max(1, node.retry.maxAttempts());
        for (int attempt = 1; attempt <= maxAttempts; attempt++) {
            node.attempts = attempt;
            try {
                result = node.run.run(ctx, node);
            } catch (Exception ex) {
                node.lastError = String.valueOf(ex.getMessage());
                if (attempt < maxAttempts) {
                    eventLog.emit("node_retry", node.id, Map.of("attempt", attempt, "max_attempts", maxAttempts, "reason", node.lastError));
                    sleep(node.retry.backoffMillis() * attempt);
                    continue;
                }
                break;
            }

            node.lastResult = result; // visible to compensate even if the exit gate rejects this attempt

            NamedPolicy failedPolicy = null;
            PolicyResult failedResult = null;
            for (NamedPolicy policy : node.exitPolicies) {
                PolicyResult pr = policy.check(ctx, node, result);
                if (!pr.passed()) {
                    failedPolicy = policy;
                    failedResult = pr;
                    break;
                }
            }

            if (failedPolicy == null) {
                ctx.setOutput(node.id, result);
                node.status = NodeStatus.SUCCEEDED;
                eventLog.emit("node_succeeded", node.id, Map.of("stage", node.stage));
                return;
            }

            ctx.recordPolicyViolation(node.id, failedPolicy.name(), failedResult.message());
            eventLog.emit("policy_violation", node.id, Map.of("policy", failedPolicy.name(), "gate", "exit", "message", failedResult.message()));
            node.lastError = "exit policy '" + failedPolicy.name() + "' failed: " + failedResult.message();
            if (attempt < maxAttempts) {
                eventLog.emit("node_retry", node.id, Map.of("attempt", attempt, "max_attempts", maxAttempts, "reason", failedResult.message()));
                sleep(node.retry.backoffMillis() * attempt);
                continue;
            }
            break;
        }

        if (node.fallback != null) {
            try {
                eventLog.emit("node_fallback_invoked", node.id, Map.of());
                result = node.fallback.run(ctx, node);
                node.lastResult = result;
                ctx.setOutput(node.id, result);
                node.status = NodeStatus.SUCCEEDED;
                eventLog.emit("node_succeeded", node.id, Map.of("stage", node.stage, "via_fallback", true));
                return;
            } catch (Exception ex) {
                node.lastError = "fallback also failed: " + ex.getMessage();
            }
        }

        node.status = NodeStatus.FAILED;
        eventLog.emit("node_failed", node.id, Map.of("stage", node.stage, "error", String.valueOf(node.lastError)));
    }

    private void rollback() {
        List<String> order = new ArrayList<>(graph.topologicalOrder());
        Collections.reverse(order);
        eventLog.emit("rollback_started", null, Map.of("order", order));
        for (String nid : order) {
            Node node = graph.nodes.get(nid);
            boolean eligible = (node.status == NodeStatus.SUCCEEDED || node.status == NodeStatus.FAILED)
                    && node.compensate != null && node.lastResult != null;
            if (eligible) {
                try {
                    node.compensate.compensate(ctx, node);
                    boolean wasSucceeded = node.status == NodeStatus.SUCCEEDED;
                    if (wasSucceeded) node.status = NodeStatus.ROLLED_BACK;
                    eventLog.emit("node_rolled_back", node.id, Map.of("was_succeeded", wasSucceeded));
                } catch (Exception ex) {
                    eventLog.emit("node_rollback_failed", node.id, Map.of("error", String.valueOf(ex.getMessage())));
                }
            }
        }
        for (Node node : graph.nodes.values()) {
            if (node.status == NodeStatus.PENDING || node.status == NodeStatus.STALE) {
                node.status = NodeStatus.SKIPPED;
            }
        }
        eventLog.emit("rollback_finished");
    }

    private void safeStop() {
        List<String> skipped = new ArrayList<>();
        for (Node node : graph.nodes.values()) {
            if (node.status == NodeStatus.PENDING || node.status == NodeStatus.STALE) {
                node.status = NodeStatus.SKIPPED;
                skipped.add(node.id);
            }
        }
        eventLog.emit("safe_stop", null, Map.of(
                "skipped_nodes", skipped,
                "note", "halted without rollback; all previously SUCCEEDED nodes remain in place as last-known-good state"
        ));
    }

    private static void sleep(long millis) {
        if (millis <= 0) return;
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
